### PR TITLE
feat(checkbox): add labelPlacement and justify props

### PR DESCRIPTION
**Pull request by amandaejohnston**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the new behavior?

Adds the `labelPlacement` and `justify` props to checkbox, along with screenshot tests. Also adds an a11y test and a fix to get that passing.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by liamdebeasi — COMMENTED** [review]
Noticing some misalignment in `core/src/components/checkbox/test/label/checkbox.e2e.ts-snapshots/checkbox-label-start-justify-start-md-rtl-Mobile-Chrome-linux.png`. Looks like there's some extra space above the checkbox?

**Comment by amandaejohnston** [comment]
@liamdebeasi That looks to just be a browser rendering quirk; the text is 19px tall with the current font choice, and the checkbox is 18px, so it has to be off-center one way or the other. I confirmed that the bounding boxes line up perfectly at the bottom edge.

File: core/src/components/checkbox/checkbox.scss [hunk 1]
```diff
@@ -158,6 +156,99 @@ input {
...
+// ---------------------------------------------
+
+:host(.checkbox-justify-space-between) .checkbox-wrapper {
+  justify-content: space-between;
+}
+
+:host(.checkbox-justify-space-between.checkbox-label-placement-start) .checkbox-icon,
+:host(.checkbox-justify-space-between.checkbox-label-placement-fixed) .checkbox-icon {
+  @include margin(0, 0, 0, auto);
+}
```
Comment (liamdebeasi) [inline core/src/components/checkbox/checkbox.scss:170]: I recommend wrapping `.checkbox-icon` in a `.native-wrapper` element like the other new form controls have. This way you can apply margins to the icon container and not worry about interfering with the SVG itself.

From there, you can get rid of these `.checkbox-icon` selectors and apply the following CSS:

```
.native-wrapper {
  display: flex;

  align-items: center;
}
```

You can look at the Toggle CSS code to see how it was done there too.

**Review by liamdebeasi — CHANGES_REQUESTED** [review]
A few things:

1. Can we add a test to make sure the label text gets truncated with ellipses properly? (in general, not just with `labePlacement="fixed"`
2. I'm seeing that an empty `<label>` element is still rendered inside of the parent `<label>` This likely left over from the legacy rendering.
3. Is `aria-label` functionality part of this PR? If so, we should add an a11y test with automated axe checks.

**Comment by amandaejohnston** [comment]
Adjusted a few things per the above 👀 (cc @sean-perkins This is ready for review for real now)

**Comment by sean-perkins** [comment]
Implementation and changes look good to me! One question, is there a reason we aren't doing `item` tests as part of this PR?

Comparing against toggle: https://github.com/ionic-team/ionic-framework/pull/26314 and I know I also ran into adding those tests for label placement on checkbox. Is that test criteria covered in existing/other tests?

**Review by liamdebeasi — APPROVED** [review]

**Review by sean-perkins — APPROVED** [review]

**Comment by amandaejohnston** [comment]
@sean-perkins Item tests are handled here: https://github.com/ionic-team/ionic-framework/pull/26471

**Pull request merged**